tests/test_digital.py

```diff
@@ -275,3 +275,53 @@ def test_U() -> None:
     assert torch.allclose(
         u(state, values), pyq.QuantumCircuit(n_qubits, u.digital_decomposition())(state, values)
     )
+
+
+def test_dm() -> None:
+    state_00 = pyq.utils.product_state("00")
```

Comment:
I add a test with |101> [here](https://github.com/pasqal-io/pyqtorch/pull/147/commits/b94394268c184c5b92d98f24aa17b5d202d27e54).

Reply:
Indeed. What I mean was maybe something along parametrising the test with a sample of bitstrings and assert the density matrix in the end. This would avoid hardcoding the resulting dm.

Reply:
What do you think of this version ?

```python

def test_dm() -> None:
    #Verification without batches 
    state_list = ['1','01','10','101','1001','10101','001111']
    for state in state_list:
        vector = product_state(state)
        assert torch.allclose(
            density_mat(vector), pyq.Projector(0, ket= state, bra= state).unitary()
            ), "The density matrix is not the projetctor."


    #Verification with batches 
    state_00_batch = product_state("00", batch_size=3)
    dm_batch = density_mat(state_00_batch)
    dm_00_batch = torch.tensor(
        [
            [
                [1.0 + 0.0j, 1.0 + 0.0j, 1.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
            ],
            [
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
            ],
            [
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
            ],
            [
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j],
            ],
        ]
    )
    assert torch.allclose(
        dm_00_batch, dm_batch
    ), "The density matrix is not the projetctor |00><00| with 3 batches."

    #Verification on random state
    state_random = random_state(n_qubits=2)
    dm_random = density_mat(state_random)
    indices_to_check = [(0, 0), (1, 1), (2, 2), (3, 3)]
    for i, j in indices_to_check:
        assert (
            dm_random[i, j, 0].imag == 0
        ), f"The imaginary part of the element at index ({i}, {j}) is not zero."

Reply:
What I had in mind is using `pytest` parametrisation functionality, see the docs [here](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html).

Reply:
What do you think of this version ?
```python
@pytest.mark.parametrize("n_qubits,batch_size",torch.randint(1,6,(8,2)))
def test_dm_batch(n_qubits: Tensor, batch_size: Tensor):
    #Test without batches:
    state = random_state(n_qubits)
    proj = torch.outer(state.flatten(),state.conj().flatten()).view(2**n_qubits,2**n_qubits,1)
    dm = density_mat(state)
    assert torch.allclose(dm,proj),"The density matrix is not the projector."

    #Test with batches:
    state_list=[]
    proj_list = [] 
    #Batches creation:
    for batch in range(batch_size): 
        #Batch state creation:
        state = random_state(n_qubits) 
        state_list.append(state)
        #Batch projector: 
        proj = torch.outer(state.flatten(),state.conj().flatten()).view(2**n_qubits,2**n_qubits,1)
        proj_list.append(proj)
    #Concatenate all the batch projectors:
    dm_proj = torch.cat(proj_list,dim=2)    
    #Concatenate the batch state to compute the density matrix   
    state_cat = torch.cat(state_list,dim=n_qubits)
    dm = density_mat(state_cat)
    assert torch.allclose(dm,dm_proj),"The density matrix is not the projector."

Reply:
Yeah something along these lines. You can probably also check the size of the dm is what you expect.